### PR TITLE
Try to stabilize our S3 integration tests

### PR DESCRIPTION
**Pull request by abrooksv**
* Rewrote our waiter to not returns strings on failures, callee code should worry about that
* Wait until head bucket works after creating a bucket to try to stabilize our tests

## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: core/src/software/aws/toolkits/core/utils/Waiter.kt [hunk 1]
```diff
@@ -3,14 +3,87 @@
...
+                }
+
+                delay(calculateBackOff(attempt))
+            }
+
+            throw WaiterTimeoutException("Condition was not met after $attempt attempts (${Duration.between(start, Instant.now()).seconds} seconds)")
+        }
+    }
+
+    private fun calculateBackOff(attempt: Int) = 250L shl min(attempt - 1, 4) // Max = 250 * 2^4 = 4_000.
```
Comment (hunterwerlla) [inline core/src/software/aws/toolkits/core/utils/Waiter.kt:77]: That's the max, but what strategy is captured here? It looks like exponential backoff? also backoffms?

File: core/src/software/aws/toolkits/core/utils/Waiter.kt [hunk 2]
```diff
@@ -3,14 +3,87 @@
...
+
+object Waiters {
+    private val LOG = getLogger<Waiters>()
+
+    /**
+     * Generic waiter method to wait for a certain AWS resource to be in a steady status until it reaches in a failed status
+     * or time out for pulling.
+     * @param T The AWS resource type for querying the status from it
+     */
+    fun <T> waitUntil(
```
Comment (hunterwerlla) [inline core/src/software/aws/toolkits/core/utils/Waiter.kt:23]: Do we want to make our waiter suspend then allow callers to runBlocking if they want?

File: core/src/software/aws/toolkits/core/utils/Waiter.kt [hunk 3]
```diff
@@ -3,14 +3,87 @@
...
+        // The success predicate based on the returned resource
+        succeedOn: (T) -> Boolean = { true },
+        // The fail predicate based on the returned resource. Return an error message if it fails, null otherwise
+        failOn: (T) -> Boolean = { false },
+        exceptionsToStopOn: Set<KClass<out Exception>> = emptySet(),
```
Comment (hunterwerlla) [inline core/src/software/aws/toolkits/core/utils/Waiter.kt:28]: it's not immediately clear what this means? shouldn't we stop on all non ignored exceptions?

File: core/src/software/aws/toolkits/core/utils/Waiter.kt (hunk 1 above)
Comment (abrooksv) [reply core/src/software/aws/toolkits/core/utils/Waiter.kt:77]: I stole it from the Java SDK test util waiter...its exponential backoff

File: core/src/software/aws/toolkits/core/utils/Waiter.kt (hunk 2 above)
Comment (abrooksv) [reply core/src/software/aws/toolkits/core/utils/Waiter.kt:23]: Not sure, puts more bloat into the call sites.... seems like the name would indicate it should always block

File: core/src/software/aws/toolkits/core/utils/Waiter.kt (hunk 3 above)
Comment (abrooksv) [reply core/src/software/aws/toolkits/core/utils/Waiter.kt:28]: Ya ill document the params

File: core/src/software/aws/toolkits/core/utils/Waiter.kt (hunk 1 above)
Comment (hunterwerlla) [reply core/src/software/aws/toolkits/core/utils/Waiter.kt:77]: Right but can we add a comment about it + that it's in MS

File: core/src/software/aws/toolkits/core/utils/Waiter.kt (hunk 3 above)
Comment (abrooksv) [reply core/src/software/aws/toolkits/core/utils/Waiter.kt:28]: And no, things like IOExceptions should be retried, its more like a if we trigger a delete we want to keep retrying until we get a ResourceNotFoundException so we should stop then

File: core/src/software/aws/toolkits/core/utils/Waiter.kt (hunk 2 above)
Comment (hunterwerlla) [reply core/src/software/aws/toolkits/core/utils/Waiter.kt:23]: it puts more bloat on non-coroutine call sites, but shouldn't this be called from coroutines like 90%+ of the time?

File: core/src/software/aws/toolkits/core/utils/Waiter.kt (hunk 3 above)
Comment (hunterwerlla) [reply core/src/software/aws/toolkits/core/utils/Waiter.kt:32]: Cool

**Pull request merged**